File: code/Algorithm/leetcode/78.子集.cpp
/*
 * @lc app=leetcode.cn id=78 lang=cpp
 *
 * [78] 子集
 *
 * https://leetcode-cn.com/problems/subsets/description/
 *
 * algorithms
 * Medium (79.04%)
 * Likes:    845
 * Dislikes: 0
 * Total Accepted:    162.5K
 * Total Submissions: 205.3K
 * Testcase Example:  '[1,2,3]'
 *
 * 给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。
 * 
 * 说明：解集不能包含重复的子集。
 * 
 * 示例:
 * 
 * 输入: nums = [1,2,3]
 * 输出:
 * [
 * ⁠[3],
 * [1],
 * [2],
 * [1,2,3],
 * [1,3],
 * [2,3],
 * [1,2],
 * []
 * ]
 * 
 */

#include <vector>
using namespace std;

// @lc code=start
class Solution
{
public:
    vector<vector<int>> res;

    void dfs(vector<int> &nums, vector<int> cur, vector<bool> flag, int len)
    {
        if (cur.size() == len)
        {
            res.push_back(cur);
            return;
        }
        for (int i = 0; i < nums.size(); i++)
        {
            if (flag[i])
            {
                flag[i] = false;
                cur.push_back(nums[i]);
                dfs(nums, cur, flag ,len);
                cur.pop_back();
            }
        }
    }

    vector<vector<int>> subsets(vector<int> &nums)
    {

        
        // 长度从1~len
        for (int i = 0; i <= nums.size(); i++)
        {
            vector<int> vec;
            vector<bool> flag(nums.size(), true);
            dfs(nums, vec, flag, i);
        }
        return res;
    }
};
// @lc code=end

File: code/JavaLab/src/main/java/algorithm/leetcode/search/78.子集.java
package algorithm.leetcode.search;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * https://leetcode-cn.com/problems/subsets/description/
 * <p>
 * 给你一个整数数组 nums ，数组中的元素 互不相同 。返回该数组所有可能的子集（幂集）。
 * <p>
 * 解集 不能 包含重复的子集。你可以按 任意顺序 返回解集。
 */
class LC78 {

    List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsets(int[] nums) {
        res.add(new ArrayList<>());
        LinkedList<Integer> list = new LinkedList<>();
        dfs(nums, nums.length, 0, list);
        return res;
    }

    public void dfs(int[] nums, int k, int step, LinkedList<Integer> list) {

        if (list.size() == k) {
            return;
        }

        for (int i = step; i < nums.length; i++) {
            list.add(nums[i]);
            res.add(new ArrayList<>(list));
            // 这里要从i开始才能去重
            dfs(nums, k, i + 1, list);
            list.removeLast();
        }
    }

}
